File: Java/src/gestorAplicacion/productosFinancieros/Ahorro.java
/* Clase Ahorro
 *
 * Es el producto financiero más simple, diferenciándose de su clase padre Cuenta al ofrecer una tasa de interés que producirá 
 * ingresos en el tiempo.
 *
 */
package gestorAplicacion.productosFinancieros;

import java.util.ArrayList;
import java.io.Serializable;
import gestorAplicacion.infoClientes.Banco;
import gestorAplicacion.infoClientes.Transaccion;
import gestorAplicacion.infoClientes.Usuario;
import gestorAplicacion.productosFinancieros.Cuenta;

public class Ahorro extends Cuenta implements Serializable{

	protected float tasaDeInteres;
	
    // El Array de clase de clientes de encarga de guardar todas las instancias de
    // Cliente para poder guardar y cargarlas en la serializacion
	private static ArrayList<Ahorro> ahorro = new ArrayList<>();

    //Constructor
	public Ahorro(int nroCuenta, Usuario titular,float saldo,Banco banco,String tipoDeCuenta,float tasaDeInteres, ArrayList<Transaccion> historial_ant) {
		super(nroCuenta,titular,saldo,banco,tipoDeCuenta);
		this.historialTransferencia = historial_ant;
		this.tasaDeInteres = tasaDeInteres;
		banco.getListaCuentas().add(this);
		titular.getCuentasAsociadas().add(this);
		
		ahorro.add(this);
	}
	
	public Ahorro(int nroCuenta, Usuario titular,float saldo,Banco banco,String tipoDeCuenta,float tasaDeInteres) {
		this(nroCuenta,titular,saldo,banco,tipoDeCuenta,tasaDeInteres,new ArrayList<Transaccion>());
		
	}
	
	
	// Verifica si el usuario cuenta con saldo para realizar la transacción. De ser así realiza los ajustes de saldo en cada cuenta,
    // crea el objeto transaccion y lo añade al historial de las cuentas involucradas.
	public boolean transferir(Cuenta cuentaDestino, float valorTransferencia) {
		if(this.saldo >= valorTransferencia) {
			this.restarSaldo(valorTransferencia);
			cuentaDestino.sumarSaldo(valorTransferencia);
			Transaccion trans = new Transaccion(this,cuentaDestino,valorTransferencia);
			historialTransferencia.add(trans);
			cuentaDestino.historialTransferencia.add(trans);
			return true;
		}
		else {
			return false;
		}
	}
	
	// Verifica si el usuario cuenta con saldo para realizar el pago de la cuota mensual. De ser así realiza los ajustes de saldo
    // en la cuenta origen y la deuda, crea el objeto transaccion y lo añade al historial de la cuenta de origen.
	public boolean transferir(Credito credito) {
		if(this.saldo >= credito.getCuotaMensual()) {
			this.restarSaldo(credito.getCuotaMensual());
			credito.setDeuda(credito.getDeuda()-credito.getCuotaMensual());
			Transaccion trans = new Transaccion(this,credito.getBanco().getNombreBanco(),credito.getCuotaMensual());
			historialTransferencia.add(trans);
			return true;
		}
		else {
			return false;
		}
	}

//	-------------------------------------- Metodos get-set --------------------------------------

    public static ArrayList<Ahorro> getAhorro() { return ahorro; }

    public static void setAhorro(ArrayList<Ahorro> ahorro) { Ahorro.ahorro = ahorro; }
	
	public void setTasaDeInteres(float tasaDeInteres) { this.tasaDeInteres = tasaDeInteres; }

	public float getTasaDeInteres() { return this.tasaDeInteres; }
	
	public ArrayList<Transaccion> getHistorialTransferencia(){ return this.historialTransferencia; }
	
	public void setHistorialTransferencia(ArrayList<Transaccion> historial){ this.historialTransferencia = historial; }

}

File: Python/src/Wunallet/capaLogica/gestorAplicacion/productosFinancieros/ahorro.py
# Clase Ahorro

# Es el producto financiero más simple, diferenciándose de su clase padre Cuenta al ofrecer una tasa de interés que producirá
# ingresos en el tiempo.

from multipledispatch import dispatch
from Wunallet.capaLogica.gestorAplicacion.infoClientes.transaccion import Transaccion
from Wunallet.capaLogica.gestorAplicacion.productosFinancieros.credito import Credito
from Wunallet.capaLogica.gestorAplicacion.productosFinancieros.cuenta import Cuenta
from Wunallet.capaLogica.gestorAplicacion.infoClientes.banco import Banco
from Wunallet.capaLogica.gestorAplicacion.infoClientes.usuario import Usuario



class Ahorro(Cuenta):
    '''El Array de clase de clientes de encarga de guardar todas las instancias de
    Cliente para poder guardar y cargarlas en la serializacion'''
    ahorro = []

    #Constructor
    def __init__(self, nroCuenta, titular, saldo, banco, tipoDeCuenta,tasaDeInteres,historial_ant=[]):
        super().__init__(nroCuenta,titular,saldo,banco,tipoDeCuenta)
        self.historialTransferencia=historial_ant
        self.tasaDeInteres=tasaDeInteres
        banco.getListaCuentas().append(self)
        titular.getCuentasAsociadas().append(self)


        Ahorro.ahorro.append(self)

    '''Verifica si el usuario cuenta con saldo para realizar la transacción. De ser así realiza los ajustes de saldo en cada cuenta,
    crea el objeto transaccion y lo añade al historial de las cuentas involucradas.'''

    @dispatch(Cuenta,float)
    def transferir(self,cuentaDestino,valorTransferencia):
        if self.saldo>= valorTransferencia:
            self.restarSaldo(valorTransferencia)
            cuentaDestino.sumarSaldo(valorTransferencia)
            trans = Transaccion(self,cuentaDestino,valorTransferencia)
            self.historialTransferencia.append(trans)
            cuentaDestino.historialTransferencia.append(trans)
            return True
        else:
            return False


    '''Verifica si el usuario cuenta con saldo para realizar el pago de la cuota mensual. De ser así realiza los ajustes de saldo
    en la cuenta origen y la deuda, crea el objeto transaccion y lo añade al historial de la cuenta de origen.'''
    @dispatch(Credito)
    def transferir(self,credito):
        if self.saldo >= credito.getCuotaMensual():
            self.restarSaldo(credito.getCuotaMensual())
            credito.setDeuda(credito.getDeuda()-credito.getCuotaMensual())
            trans = Transaccion(self,credito.getBanco().getNombreBanco(),credito.getCuotaMensual())
            self.historialTransferencia.append(trans)
            return True
        else:
            return False

    #-------------------------------------- Metodos get-set --------------------------------------
    @classmethod
    def getAhorro(cls): return Ahorro.ahorro

    @classmethod
    def setAhorro(cls, ahorro):  Ahorro.ahorro = ahorro
	
    def  setTasaDeInteres(self, tasaDeInteres) : self.tasaDeInteres = tasaDeInteres

    def  getTasaDeInteres(self) : return self.tasaDeInteres
	
    def  getHistorialTransferencia(self): return self.historialTransferencia
	
    def  setHistorialTransferencia(self,historial): self.historialTransferencia = historial


